File: 백준/Gold/2001. 보석 줍기/보석 줍기.py
import sys
from collections import deque


def input():
    return sys.stdin.readline().rstrip()


def solution() -> int:
    n, m, K = map(int, input().split())

    jewel = [-1] * (n + 1)
    for i in range(K):
        jewel[int(input())] = i

    graph = [[] for _ in range(n + 1)]
    for i in range(m):
        a, b, c = map(int, input().split())
        graph[a].append((b, c))
        graph[b].append((a, c))

    visit = [[False for col in range(1 << K)] for row in range(n + 1)]

    dq = deque()
    dq.append((1, 0, 0))
    visit[1][0] = True

    ans = 0
    while dq:
        node, mask, numJewel = dq.popleft()

        if node == 1 and numJewel > ans:
            ans = numJewel

        for nextNode, limit in graph[node]:
            if numJewel <= limit:
                if not visit[nextNode][mask]:
                    dq.append((nextNode, mask, numJewel))
                    visit[nextNode][mask] = True

                j = jewel[nextNode]
                if j != -1 and not mask & 1 << j:
                    if not visit[nextNode][mask | 1 << j]:
                        dq.append((nextNode, mask | 1 << j, numJewel + 1))
                        visit[nextNode][mask] = True
    return ans


if __name__ == '__main__':
    print(solution())

File: 백준/Gold/2001. 보석 줍기/보석 줍기.cc
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
using namespace std;
typedef pair<int, int> pi;
typedef tuple<int, int, int> ti;

vector<int> jewel;
vector<vector<pi>> graph;
vector<vector<bool>> visit;
queue<ti> q;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);


    int n, m, K; cin >> n >> m >> K;

    jewel.resize(n + 1, -1);

    for (int i = 0; i < K; i++) {
        int nd; cin >> nd;

        jewel[nd] = i;
    }

    graph.resize(n + 1, vector<pi>());
    visit.resize(n + 1, vector<bool>(1 << K, false));
    for (int i = 0; i < m; i++) {
        int a, b, c; cin >> a >> b >> c;

        graph[a].emplace_back(b, c);
        graph[b].emplace_back(a, c);
    }

    q.emplace(1, 0, 0);
    visit[1][0] = true;
    int ans = 0;
    int node, mask, numJewel;
    while (!q.empty()) {
        tie(node, mask, numJewel) = q.front();
        q.pop();

        if (node == 1 && numJewel > ans) {
            ans = numJewel;
        }

        for (auto &[nextNode, limit] : graph[node]) {
            if (numJewel <= limit && !visit[nextNode][mask]) {
                q.emplace(nextNode, mask, numJewel);
                visit[nextNode][mask] = true;

                int j = jewel[nextNode];
                if (j != -1 && !(mask & 1 << j) && !visit[nextNode][mask | 1 << j]) {
                    q.emplace(nextNode, mask | 1 << j, numJewel + 1);
                    visit[nextNode][mask | 1 << j] = true;
                }
            }
        }
    }
    cout << ans;
    return 0;
}